### PR TITLE
[ROCKETMQ-204]-when all brokers is offline, client still attempts to send heartbeat

### PR DESCRIPTION
**Pull request by Jaskey**
JIRA: https://issues.apache.org/jira/browse/ROCKETMQ-204?jql=project%20%3D%20ROCKETMQ

### CODE REVIEW

File: client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPullConsumerImpl.java [hunk 1]
```diff
@@ -324,11 +324,16 @@ public void persistConsumerOffset() {
...
+        if (info != null) {
+            Map<String, SubscriptionData> subTable = this.rebalanceImpl.getSubscriptionInner();
+            if (subTable != null) {
+                if (subTable.containsKey(topic)) {
+                    this.rebalanceImpl.getTopicSubscribeInfoTable().put(topic, info);
+                }
             }
+        } else {
+            Set<MessageQueue> prev = this.rebalanceImpl.getTopicSubscribeInfoTable().remove(topic);
+            log.info("instanceName={}, group={}, topicSubscribeInfoTable of topic {} is remove, {}, prev = {}", defaultMQPullConsumer.getInstanceName(), defaultMQPullConsumer.getConsumerGroup(), topic, prev);
```
Comment (lizhanhui) [inline client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPullConsumerImpl.java]: is remove --> is removed

File: client/src/main/java/org/apache/rocketmq/client/impl/factory/MQClientInstance.java [hunk 2]
```diff
@@ -621,34 +621,23 @@ public boolean updateTopicRouteInfoFromNameServer(final String topic, boolean is
...
-                                }
+                                updateSubInfoTable(topic, subscribeInfo);
                             }
                             log.info("topicRouteTable.put. Topic = {}, TopicRouteData[{}]", topic, cloneTopicRouteData);
                             this.topicRouteTable.put(topic, cloneTopicRouteData);
                             return true;
                         }
                     } else {
-                        log.warn("updateTopicRouteInfoFromNameServer, getTopicRouteInfoFromNameServer return null, Topic: {}", topic);
+                        log.warn("updateTopicRouteInfoFromNameServer, getTopicRouteInfoFromNameServer return null. Topic: {}", topic);
```
Comment (vsair) [inline client/src/main/java/org/apache/rocketmq/client/impl/factory/MQClientInstance.java]: IMO,  API(getTopicRouteInfoFromNameServer) will throw exception when not found topic route, so the logic will never enter 'else', am I right?

Comment (Jaskey) [inline client/src/main/java/org/apache/rocketmq/client/impl/factory/MQClientInstance.java]: It seems you are right, this design is confusing , maybe we change the logic into returns null?

**Comment by Jaskey** [comment]
@vsair @lizhanhui please review the updated pr

**Comment by lizhanhui** [comment]
@Jaskey I'll check it today.

**Comment by Jaskey** [comment]
@lizhanhui @vsair @zhouxinyu 

any opinions?

**Comment by vongosling** [comment]
@Jaskey I will close the pr, if you happened to the same question, please let me know.

**Pull request closed without merging**